src/librustc_resolve/diagnostics.rs

```diff
@@ -444,6 +444,52 @@ impl SomeTrait for Foo { // ok!
 ```
 "##,
 
+E0406: r##"
+A function is referring to an associated type which hasn't been declared in
+the trait. Example of erroneous code:
+
+```compile_fail
+trait Foo {
+    type Bar;
+
+    // error: function signature contains a reference to `Self::Baz`, but
+    //        `Baz` hasn't been declared as an associated type of the trait
+    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
+}
+```
+
+The use of 'associated types' allows us declare variables which might be used
```

Comment:
Soory, I'll try to explain more clearly. So in here, what's expected is something which helps the user understands the error. Easy example:

```rust
let mut x: u32 = 0;
x = "salut !";
```

It'll fail because the expected type doesn't correspond to the one we put in `x`. So instead of explaining what is a variable declaration, it'd be nice to point out that the two types __must__ be the same. In this case, I'd write:

"The type of a variable cannot change once declared. If you want the type of `x` to change, you must redeclare it but you'll lose its previous content. Otherwise you can just store the other data into a new variable."

And then you put examples which show it. Did it help?